### PR TITLE
Adding support for `hidden_states` and `attentions` in unbatching support.

### LINKED ISSUE #14414: Pipelines fails with IndexError using Bert model with outputs and batch size >= 16
## Environment info

- `transformers` version: 4.13.0.dev0
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.12
- PyTorch version (GPU?): 1.10.0+cu111 (True)
- Tensorflow version (GPU?): 2.7.0 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help
@Narsil 

Library:
- Pipelines: @Narsil

## Information

Model I am using (Bert, XLNet ...): [FinBert](https://huggingface.co/yiyanghkust/finbert-tone)

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

I came across this issue when setting  `output_hidden_states=True` during instantiation of a pretrained model inorder to obtain the inferred CLS sentence embeddings using the following approach:

```
from transformers.pipelines.text_classification import TextClassificationPipeline
class FinBertSentimentClassificationPipeline(TextClassificationPipeline):
  def postprocess(self, model_outputs, function_to_apply=None, return_all_scores=False):
    prediction = super().postprocess(model_outputs, function_to_apply, return_all_scores)
    prediction['last_hidden_layer']= model_outputs.hidden_states[0][0][0]
    return prediction


def custom_pipeline(task,model,tokenizer,**kwargs):
  kwargs['tokenizer'] = tokenizer
  kwargs['feature_extractor'] = None
  return FinBertSentimentClassificationPipeline(model=model,framework='pt',task=task,**kwargs)
```

Where I override the postprocess function to also return the last hidden state layer. However, the issue occurs in the pipeline code itself when using a batch size equal to or larger than 16. 
## To reproduce

1. `!pip install git+https://github.com/huggingface/transformers.git`
2. 
```
from transformers import BertTokenizerFast, BertForSequenceClassification
from transformers import pipeline
finbert = BertForSequenceClassification.from_pretrained('yiyanghkust/finbert-tone',num_labels=3, output_hidden_states=True)
tokenizer = BertTokenizerFast.from_pretrained('yiyanghkust/finbert-tone')

nlp = pipeline("sentiment-analysis", model=finbert, tokenizer=tokenizer, device=0)

varying_length_sentences = ["there is a shortage of capital, and we need extra financing "*5,  
             "growth is strong and we have plenty of liquidity ", 
             "there are doubts about our finances" * 10, 
             "profits are flat",
             "profits are flat "*30]*1000
similar_length_sentences = ["there is a shortage",  
             "growth is strong ", 
             "there are doubts", 
             "profits are flat"]*1000
```
3. `results = nlp(similar_length_sentences, batch_size=16, num_workers=2)`
See the [Colab Notebook](https://colab.research.google.com/drive/1SlntHYK-F8my84rUiS1xuqQN4v7nBgEN?usp=sharing) for reference.

Running step 3 produces the following error:
```
/usr/local/lib/python3.7/dist-packages/transformers/pipelines/base.py in loader_batch_item(self)
    750                     if k == "past_key_values":
    751                         continue
--> 752                     if isinstance(element[self._loader_batch_index], torch.Tensor):
    753                         loader_batched[k] = element[self._loader_batch_index].unsqueeze(0)
    754                     elif isinstance(element[self._loader_batch_index], np.ndarray):

IndexError: tuple index out of range
```

Executing the pipeline with batch sizes smaller than 16 seem to work (see colab notebook).

## Expected behavior

Pipeline runs successfully with any batch size using a model loaded to output hidden states and attention.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes #14414

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/test_pipelines_common.py [hunk 1]
```diff
@@ -322,6 +323,19 @@ def data(n: int):
             results.append(out)
         self.assertEqual(len(results), 10)
 
+    @require_torch
+    def test_unbatch_attentions_hidden_states(self):
+        model = DistilBertForSequenceClassification.from_pretrained(
+            "Narsil/tiny-distilbert-sequence-classification", output_hidden_states=True, output_attentions=True
```
Comment (LysandreJik) [inline tests/test_pipelines_common.py:329]: Can this leverage the one from `hf-internal-testing`, or is there a reason you favor this one?

**Review by LysandreJik — APPROVED** [review]
Cool, thanks @Narsil

File: tests/test_pipelines_common.py (hunk 1 above)
Comment (Narsil) [reply tests/test_pipelines_common.py:329]: There is not `tiny-distilbert` in `hf-internal-testing`, only the random one.
Happy to upload it if you want (as well as all the other `Narsil` specific tests that probably all belong in `hf-internal-testing`.

For instance this one was just copied from the `text-classification` tests.

Comment (LysandreJik) [reply tests/test_pipelines_common.py:329]: Right, but the random has random weights but actually has the weights so it wouldn't be initialized randomly every time you load it in memory.

It shouldn't be an issue here given that you're only checking the length of outputs, right?

Comment (Narsil) [reply tests/test_pipelines_common.py:329]: Well I am using these also in other tests, are you ok if I merge this PR, then open another PR to remove `Narsil/` models from a bit everywhere (some check actual output values)

**Pull request merged**